### ISSUE TITLE
Question: Why are Homebrew and Python 3 required ?

### ISSUE THREAD

**Issue by section83** [Kevin-De-Koninck/Get-It#15]
Just wondering about the dependencies:

Why is Python 3 required ? youtube-dl works perfectly with default Python in macOS installation (v2.7 on my iMac with 10.13.6).  I chose to run the Python install, got the "Please Wait" dialog, but nothing happened.  Clicked the advance button and got message that everything was ready to go, clicked on the refresh button then closed the window (manually) but, main dialog still says dependencies are missing, as does prefs file.

Ditto Homebrew, why ?  I have Homebrew installed and Get.It sees it which is OK.  I have youtube-dl, FFmpeg and FFprobe installed but Get.It doesn't find them.  They are in the default location /usr/local/bin/.

Cheers.

**Comment by Kevin-De-Koninck** [comment]
Late to the party... Broke my hand last night so now I got the time! I'm sorry for the long wait. I'll be fixing most issues in the coming days (as many as I can).

**Comment by Kevin-De-Koninck** [comment]
Hi @section83 

Probably a bug and such :) I removed some code that checks if the required software is installed and removed the installer wizard, users now have to install the dependencies (see readme) or use the new button in the options view).

regarding python 3: I actually don't remember it anymore. maybe it had something to do with `pycrypt`. I should investigate it in the near future... 

Homebrew was a requirement because I installed youtube-dl and others via brew.


I'll upload a new version soon and bump this issue again so you can test it 👍

**Comment by Kevin-De-Koninck** [comment]
New version available here:
https://github.com/Kevin-De-Koninck/Get-It/releases/tag/v0.7

@section83 can you verify?

**Comment by section83** [comment]
Kevin, thanks for this.  I should fess-up that I have also developed a youtube-dl GUI front end (https://forum.videohelp.com/threads/384566-MacYTDL-youtube-dl-GUI-for-Apple-Macs).  I made the mistake of developing MacYTDL in AppleScript.  That's limited the functionality (no progress bar, modal alert dialog only etc.).  I have been thinking of learning Swift to do something better.  You've progressed a lot further so, I'll avoid redeveloping MacYTDL – but, I might be able to help out if I make progress with learning Swift.  If I do learn Swift, it'll be Swift 5.  Would it be better to upgrade Get-It to Swift 5 ?

Anyway, I've tested the updated Get-It and have some suggestions (some you may already have in your plan):

- I'd add an option to set the downloads folder – I nearly always need to save downloads somewhere other than the macOS downloads folder (eg. external volume).

- Where are preferences stored ?

- Set a default video/audio download format – the default video format doesn't display in the preferences dialog - what if mp3 is not available for audio ? - I recommend just going with ytdl default (best:best).

- "Install/Update software" should report back something - even if it's just an "everything is up to date".

- There was an error with this URL (https://iview.abc.net.au/show/countdown-to-glory/series/1/video/LR1824V002S00). But, that site is supported by ytdl.  Another video seemed to download but, I found that the audio had downloaded but not the video.  That has happened with all download attempts.  I suspect the lack of a default video format showing in the preferences dialog is related.

- I suggest saving ytdl responses to a log file - The error message on that failed video didn't stay visible long enough for me to read what it said.  When I did get to read the error, the message didn't help me. ....  –  ....  OK, I found the log file - it will help users to have separate log files for each download - eventually, you might include an open logs function and, of course, a delete all logs function.
- "Gathering information" is displayed when the video is being downloaded.  But, it's actually downloading.  Are you doing a simulate before the download ?
- Can't open download folder while video is downloading.  Can you spawn a process for downloads so that users can change settings and start another download ? [In MacYTDL I had to spawn an _osascript_ process in background which is very ugly].
- A "Download finished" message would be nice.
- Need a user guide - for me, writing the user guide from the very beginning helped with developing - also, I've found that many users just don't understand a lot of things unless they are explained in text format.
- Need an uninstaller.
- The code signature didn't seem to work - I still had to override GateKeeper.
- Know nothing about Swift but, is it necessary to pack all those Swift frameworks into the bundle ?

A little matter is the name "Get-It".  I made the mistake of searching for "Get.It" and got a totally unrelated web site.  So, I had to think hard to find Get-It.

Cheers.

**Comment by Kevin-De-Koninck** [comment]
> Kevin, thanks for this. I should fess-up that I have also developed a youtube-dl GUI front end (https://forum.videohelp.com/threads/384566-MacYTDL-youtube-dl-GUI-for-Apple-Macs). I made the mistake of developing MacYTDL in AppleScript. That's limited the functionality (no progress bar, modal alert dialog only etc.). I have been thinking of learning Swift to do something better. You've progressed a lot further so, I'll avoid redeveloping MacYTDL – but, I might be able to help out if I make progress with learning Swift. If I do learn Swift, it'll be Swift 5. Would it be better to upgrade Get-It to Swift 5 ?

I love to see that people are developing too! Never even tought of AppleScript so that's pretty nice!
I recently updated Get It to Swift 5 ([commit](https://github.com/Kevin-De-Koninck/Get-It/commit/bd68ecfbabbe72d45008e1f2eb24edb6d41d51f8)). I'd advise to learn the latest language at once (no need to learn old stuff).
Actually I developed get-it because I wanted to learn how to develop macOS apps (and my gf wanted a youtube downloader, no CLI). If you want to practive you swift skills, go ahead on Get It if you want. it's far from perfect (haven't really had the time to deepen myself in optimising it... So... if you're really into it, I could add you to the repo so you can push directly into develop (or your own brnach).

> I'd add an option to set the downloads folder – I nearly always need to save downloads somewhere other than the macOS downloads folder (eg. external volume).

If you click the folder icon in the general settings, you'll be able to do that.

> Where are preferences stored ?

I create a dictionary and store it in user defaults ([code](https://github.com/Kevin-De-Koninck/Get-It/blob/master/Get%20It/OptionsViewController.swift#L170))
```
        UserDefaults.standard.set(settingsDict, forKey: SETTINGS_KEY)
        UserDefaults.standard.synchronize()
```
some info: https://www.hackingwithswift.com/read/12/2/reading-and-writing-basics-userdefaults

> Set a default video/audio download format – the default video format doesn't display in the preferences dialog - what if mp3 is not available for audio ? - I recommend just going with ytdl default (best:best).

back in the old days, I made the defaults this way that my GF could easily download MP3's from youtube. She couldn't care about the quality and such thus the defaults are the best for this job (easy to restore for her). But then get it got a bit more popular so these settings aren't the best defaults. You can create a seperate issue for this so that this can be tracked more easily.

This is explained in the README (should be altered too then): https://github.com/Kevin-De-Koninck/Get-It#about
`The default settings will download the audio from a video, convert it to an MP3 and save it to you downloads folder. This was, in my opinion, the mostly used setting.`

> "Install/Update software" should report back something - even if it's just an "everything is up to date".

Correct. This can be added. Will see what I can do.
1. keep a synchronous action, and report at the end
2. make it asynchronous and report during the job (kinda like downloading dialog)

I'll (or you) make a seperate issue out of this

> There was an error with this URL (https://iview.abc.net.au/show/countdown-to-glory/series/1/video/LR1824V002S00). But, that site is supported by ytdl. Another video seemed to download but, I found that the audio had downloaded but not the video. That has happened with all download attempts. I suspect the lack of a default video format showing in the preferences dialog is related.

Easiest to do is to:
```
cat /tmp/getit_logs
```

I think that the youtube-dl command will be saved as well alongside with the programs output. that simplifies the debugging. If the issue persists, please create a separate issue (easier to track thingies)

> I suggest saving ytdl responses to a log file - The error message on that failed video didn't stay visible long enough for me to read what it said. When I did get to read the error, the message didn't help me. .... – .... OK, I found the log file - it will help users to have separate log files for each download - eventually, you might include an open logs function and, of course, a delete all logs function.

correct: `/tmp/getit_logs`
The logging system is really simple so it could be improved but it was added to easily debug problems that other people had.
as again: create a seperate issue if it could be improved :) 

I kept it 'behind a user's back' because most people that I know that use Get It, don't really care about the output

> "Gathering information" is displayed when the video is being downloaded. But, it's actually downloading. Are you doing a simulate before the download ?

might be some issues with the asynchronous download (needed to be able to display information at run time). Could be worth to take a look at it.

> A "Download finished" message would be nice.

can be added here: https://github.com/Kevin-De-Koninck/Get-It/blob/master/Get%20It/ViewController.swift#L215

> Can't open download folder while video is downloading. Can you spawn a process for downloads so that users can change settings and start another download ? [In MacYTDL I had to spawn an osascript process in background which is very ugly].

I'm not a big fan of this. the more you run in parallel, the more issues will occur... it is possible to not disable the open downloads folder button during download, here by removing the line: https://github.com/Kevin-De-Koninck/Get-It/blob/master/Get%20It/ViewController.swift#L62

> Need a user guide - for me, writing the user guide from the very beginning helped with developing - also, I've found that many users just don't understand a lot of things unless they are explained in text format.

For me this project was more a 'software that only I will use' kind of project. It could be useful to add more tooltips to all gui fields (hover your mouse over the text field for your urls to get an example).

> Need an uninstaller.

Remove the app from the applications folder.
It's tricky to remove all dependencies: what if a user had youtube-dl but only wanted to try this GUI. But he tiught it didn't meet his expectations so he clicks 'uninstall' hoping the program will be uninstalled but is left with a broken machine (no ffmpeg, no yuotube-dl, no brew?).

It might be useful to add the commands in the readme on how to delete all the dependencies so that users can choose what they want to uninstall, but an uninstall script will be tricky.

> The code signature didn't seem to work - I still had to override GateKeeper.

Probably, I haven't signed it.

> Know nothing about Swift but, is it necessary to pack all those Swift frameworks into the bundle ?

No idea what you mean by this, but I use some cocoapods (extra frameworks) that are used for the toggle switches etc. Those are needed yes

> A little matter is the name "Get-It". I made the mistake of searching for "Get.It" and got a totally unrelated web site. So, I had to think hard to find Get-It.

Yes, not the best name but it fits in with sub it and clean it (currently named as clean-me)). Haven't done any SEO so I can believe that it's hard to find 😋 


--

Again, if you want to contribute, please do. Don't be afraid to do a PR. Download the project and start coding 😋 

and for all issues or improvements mentioned above, make seperate issues so that it can be tracked one by one.

Kr, 
Kevin

**Comment by section83** [comment]
Kevin, many thanks for all the responses and invite.  I need to learn a lot about Swift and XCode before having anything useful to contribute.  Also my iMac is in repairs (cooked GPU) and I'm working on an early 2003 15" flat panel running OS X 10.5 - which is incompatible with GitHub.  But, I'll keep in touch.

```Never even thought of AppleScript so that's pretty nice!```

With AppleScript I was able to get a lot of functionality built quickly.  But, it would have been impossible without an ASOC script library from Shane Stanley which provided dialogs (https://www.macosxautomation.com/applescript/apps/Script_Libs.html).  Shane helped a lot in getting it working.  Even so, I can't recommend AS for a GUI-heavy project.

```what if a user had youtube-dl but only wanted to try this GUI```

I hadn't thought of that.  With MacYTDL's uninstaller, everything is moved to Trash, not killed.  So, the user can move back tools they want to keep.  I think ffmpeg would be more needed than youtube-dl.  It also trashes the script library, web browser service, AtomicParsley and preferences folder (which stores log files too).  Luckily, no-one has complained so far.

Cheers.

Garry

**Comment by Kevin-De-Koninck** [comment]
I'll close this issue for now. open as many issues as you want for the items you have described above. yhanks for your input and as soon as you're ready do a PR, that way you'll learn the fastest 👍

**Comment by section83** [comment]
Good oh.  I've just borrowed my first _Swift_ text book.  Still don't have my iMac back from repairs so, might be a while before making contact again.
